bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/ScanAndCompareGarbageCollector.java

```diff
@@ -102,6 +165,68 @@ public void gc(GarbageCleaner garbageCleaner) {
             LOG.warn("Exception when iterating over the metadata {}", e);
         }
     }
-}
 
+    private Set<Long> removeOverReplicatedledgers(Set<Long> bkActiveledgers, final GarbageCleaner garbageCleaner)
+            throws InterruptedException, KeeperException {
+        final Set<Long> overReplicatedLedgers = Sets.newHashSet();
+        final Semaphore semaphore = new Semaphore(MAX_CONCURRENT_ZK_REQUESTS);
+        final CountDownLatch latch = new CountDownLatch(bkActiveledgers.size());
+        for (final Long ledgerId : bkActiveledgers) {
+            try {
+                // check if the ledger is being replicated already by the replication worker
+                if (ZkLedgerUnderreplicationManager.isLedgerBeingReplicated(zk, zkLedgersRootPath, ledgerId)) {
+                    latch.countDown();
+                    continue;
+                }
+                // we try to acquire the underreplicated ledger lock to not let the bookie replicate the ledger that is
+                // already being checked for deletion, since that might change the ledger ensemble to include the
+                // current bookie again and, in that case, we cannot remove the ledger from local storage
+                ZkLedgerUnderreplicationManager.acquireUnderreplicatedLedgerLock(zk, zkLedgersRootPath, ledgerId);
+                semaphore.acquire();
+                ledgerManager.readLedgerMetadata(ledgerId, new GenericCallback<LedgerMetadata>() {
+
+                    @Override
+                    public void operationComplete(int rc, LedgerMetadata ledgerMetadata) {
+                        if (rc == BKException.Code.OK) {
+                            // do not delete a ledger that is not closed, since the ensemble might change again and
+                            // include the current bookie while we are deleting it
+                            if (!ledgerMetadata.isClosed()) {
+                                release();
+                                return;
+                            }
+                            SortedMap<Long, ArrayList<BookieSocketAddress>> ensembles = ledgerMetadata.getEnsembles();
+                            for (ArrayList<BookieSocketAddress> ensemble : ensembles.values()) {
+                                // check if this bookie is supposed to have this ledger
+                                if (ensemble.contains(selfBookieAddress)) {
+                                    release();
+                                    return;
+                                }
+                            }
+                            // this bookie is not supposed to have this ledger, thus we can delete this ledger now
+                            overReplicatedLedgers.add(ledgerId);
```

Comment:
The check above at line 198 is verifying that the current bookie is not in any of the ensembles for the ledger. 

I think it should be safe to delete at this point (provided that the ledger metadata doesn't change again while we doing it, and for that we wait for the ledger to be closed and we acquire the auto-replication lock for the ledger). 

It's true that if the ledger is currently under-replicated, we might be end up deleting the last copy of some entries, but since this bookie is not in the ZK metadata, the entries would be anyway unreachable for the clients.